### PR TITLE
support explode outer

### PR DESCRIPTION
**Pull request by sperlingxx**
Signed-off-by: sperlingxx <lovedreamf@gmail.com>
Current PR is to support explode outer (with/without position) in spark-rapids plugin.

P.S. This PR is marked as W.I.P. because it won't pass the tests until [the cuDF fix](https://github.com/rapidsai/cudf/pull/7754) take effect.

### CODE REVIEW

File: integration_tests/src/main/python/generate_expr_test.py [hunk 1]
```diff
@@ -108,3 +129,24 @@ def test_posexplode_nested_array_data(spark_tmp_path, data_gen):
...
         conf=conf_to_enforce_split_input)
+
+#sort locally because of https://github.com/NVIDIA/spark-rapids/issues/84
+# After 3.1.0 is the min spark version we can drop this
+@ignore_order(local=True)
+@pytest.mark.parametrize('data_gen', all_gen, ids=idfn)
+def test_posexplode_outer_array_data(spark_tmp_path, data_gen):
+    data_gen = [int_gen, ArrayGen(data_gen)]
+    assert_gpu_and_cpu_are_equal_collect(
+        lambda spark: two_col_df(spark, *data_gen).selectExpr('a', 'posexplode(b)'),
```
Comment (revans2) [inline integration_tests/src/main/python/generate_expr_test.py:140]: I think this needs to be `posexplode_outer(b)`

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 2]
```diff
@@ -2405,6 +2405,7 @@ object GpuOverrides {
           TypeSig.commonCudfTypes + TypeSig.DECIMAL + TypeSig.NULL + TypeSig.ARRAY),
         (TypeSig.ARRAY + TypeSig.MAP).nested(TypeSig.all)),
       (a, conf, p, r) => new GeneratorExprMeta[Explode](a, conf, p, r) {
+        override val supportOuter: Boolean = true
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2408]: The description needs to be updated now too. It says that outer is not supported.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala [hunk 3]
```diff
@@ -2421,6 +2422,7 @@ object GpuOverrides {
         TypeSig.ARRAY.nested(
           TypeSig.commonCudfTypes + TypeSig.DECIMAL + TypeSig.NULL + TypeSig.ARRAY)),
       (a, conf, p, r) => new GeneratorExprMeta[PosExplode](a, conf, p, r) {
+        override val supportOuter: Boolean = true
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2425]: Here too the description needs to be updated.

**Review by revans2 — COMMENTED** [review]
Overall it is looking really good.

File: integration_tests/src/main/python/generate_expr_test.py (hunk 1 above)
Comment (sperlingxx) [reply integration_tests/src/main/python/generate_expr_test.py]: My bad. Fixed!

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala (hunk 2 above)
Comment (sperlingxx) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2407]: Fixed! Thx

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala (hunk 3 above)
Comment (sperlingxx) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuOverrides.scala:2423]: Fixed! Thx

**Review by revans2 — COMMENTED** [review]
Looks good.  The docs still need to be re-generated and added to the PR.

**Comment by sperlingxx** [comment]
build

**Comment by jlowe** [comment]
Tests are failing, possibly because of https://github.com/rapidsai/cudf/issues/7787.  If that's the case, those tests should be xfailed with a pointer to that cudf issue in the xfail description.

**Comment by revans2** [comment]
@sperlingxx this code here was upmerged to 0.6, but this is targeted to 0.5.  Also there is a separate new PR for 0.6 #2215 for this same branch.  Should we close this one?

**Comment by sperlingxx** [comment]
> @sperlingxx this code here was upmerged to 0.6, but this is targeted to 0.5. Also there is a separate new PR for 0.6 #2215 for this same branch. Should we close this one?

Yes, this PR should be closed. I am sorry that I forgot to close it.

**Pull request closed without merging**